### ISSUE TITLE
How to shuffle & optimize Hugging Face datasets for LLM pre-training with StreamingDataset ?

### ISSUE THREAD

**Issue by pietrolesci** [Lightning-AI/litdata#696]
### Description
When using `StreamingDataset` with `shuffle=True`, accessing individual samples (e.g., `dataset[0]`) returns the same sample every time, despite shuffle being enabled. This makes it impossible to verify shuffling is working or to access random samples directly from the dataset.

### Environment
- LitData version: 0.2.52
- Python version: 3.11.10

### Minimal Reproduction Code
```python
from pathlib import Path
import litdata as ld
from litdata.streaming.item_loader import ParquetLoader

INDEX_PATH = Path(".litdata_index")
CACHE_DIR = Path(".litdata_cache")

# Define the Hugging Face dataset URI
hf_dataset_uri = "hf://datasets/pietrolesci/finewebedu-20B/bpe32000minipile"
INDEX_PATH.mkdir(exist_ok=True, parents=True)
ld.index_parquet_dataset(hf_dataset_uri, cache_dir=INDEX_PATH)

dataset = ld.StreamingDataset(
    hf_dataset_uri, 
    item_loader=ParquetLoader(), 
    cache_dir=CACHE_DIR, 
    index_path=INDEX_PATH,
    shuffle=True,
    drop_last=True,
    max_pre_download=5,
)

# This returns the same sample every time
print("Sample 0:", dataset[0])
print("Sample 0 again:", dataset[0])
```

For context, I ran into this issue while trying to use LitData to transform a Hugging Face dataset into a packed dataset for LLM pretraining. To simplify, I started from a tokenized dataset and wanted to:

- Shuffle the documents
- Pack them into equal-length sequences
- Shuffle the sequences

Any tips about this would be really helpful. Thanks a lot in advance for your attention!

**Comment by bhimrazy** [comment]
For optimizing the tokenized dataset, you may also want to check out the **LLM Pre-training** under Features section in the README. Since your dataset is already tokenized, you likely don’t need to apply another tokenizer — you can just return the tokenized ids from the optimize fn.

Please feel free to ask any questions here if anything’s unclear.

**Comment by bhimrazy** [comment]
That’s great @pietrolesci 🎉, glad to hear you got it working!

On performance: everything looks solid — the only thing I’d point out is the chunk size. For local storage your current `5GB` chunks may still work fine, but they’re on the larger side. If you ever plan to stream from cloud/object storage, such large chunks could hurt performance (latency, retries, etc.). We generally recommend keeping chunk sizes ≤500MB, with **64MB being the sweet spot** for streaming workloads.

Also, about the warning pls feel free to ignore it. 

cc: @tchaton in case there are any extra tips worth adding 🙌

**Comment by pietrolesci** [comment]
Thanks a lot for your feedback and super useful discussion! 
I guess the few remaining questions I have are:
- why indexing wildcard URIs does not work (e.g., `ld.index_parquet_dataset(train_uri)`)
- whether it is possible to resume a training run (previously I was handling this with the `StatefulDataloader`, which is now complaining when I use ld.StreamingDataloader` -> what is the suggested way to do this?

**Comment by bhimrazy** [comment]
Thanks for the great discussion @pietrolesci 🙌. Glad we could get everything clarified here. Closing this issue now, but feel free to re-open or start a new one if anything else comes up.

**Comment by pietrolesci** [comment]
Thanks for your help @bhimrazy!
